### PR TITLE
[mono] Implement AsAny marshalling for simple arrays

### PR DESCRIPTION
**Pull request by alexischr**
Simple here means of rank 1, and with a blittable value type as element type.

Contributes to https://github.com/dotnet/runtime/issues/35225

### CODE REVIEW

File: src/mono/mono/metadata/marshal.c [hunk 1]
```diff
@@ -6264,6 +6264,31 @@ mono_marshal_asany_impl (MonoObjectHandle o, MonoMarshalNative string_encoding,
...
+				break;
+
+		if (mono_class_is_auto_layout (eklass))
+			break;
+
+		if (m_class_is_valuetype (eklass) && (mono_class_is_explicit_layout (eklass) || m_class_is_blittable (eklass) || m_class_is_enumtype (eklass)))
+			return (arr->vector);
+
+		if (eklass == mono_get_char_class ()) {
+			char *res =  mono_utf16_to_utf8 ((mono_unichar2 *) arr->vector, arr->max_length, error);
```
Comment (alexischr) [inline src/mono/mono/metadata/marshal.c:6286]:
```suggestion
			char *res =  mono_utf16_to_utf8 ((mono_unichar2 *)arr->vector, arr->max_length, error);
```

**Comment by vargaz** [comment]
Looks ok, are there any tests for this ?

**Comment by alexischr** [comment]
@vargaz There is one test program (`AsAnyTest.cs/cpp`) and it fixes all the cases with simple arrays, but there are still failures .See https://github.com/dotnet/runtime/issues/35225#issuecomment-622203787

**Comment by alexischr** [comment]
@vargaz Can you approve?

**Comment by alexischr** [comment]
@monojenkins build failed

**Comment by SamMonoRT** [comment]
@vargaz @lambdageek  - please review, thank you.

File: src/mono/mono/metadata/marshal.c [hunk 2]
```diff
@@ -6264,6 +6264,31 @@ mono_marshal_asany_impl (MonoObjectHandle o, MonoMarshalNative string_encoding,
...
+	case MONO_TYPE_SZARRAY: {
+		MonoClass *klass = t->data.klass;
+		MonoClass *eklass = m_class_get_element_class (klass);
+		MonoArray *arr = (MonoArray *) MONO_HANDLE_RAW (o);
+
+		if (m_class_get_rank (klass) > 1)
+			break;
+
+		if (arr->bounds)
+			if (arr->bounds->lower_bound != 0)
```
Comment (CoffeeFlux) [inline src/mono/mono/metadata/marshal.c:6276]: Is this a permanent restriction, or just one you're putting in place for now? Not sure how CoreCLR handles non-0 lower bounds.

File: src/mono/mono/metadata/marshal.c [hunk 3]
```diff
@@ -6264,6 +6264,31 @@ mono_marshal_asany_impl (MonoObjectHandle o, MonoMarshalNative string_encoding,
...
 	}
+	case MONO_TYPE_SZARRAY: {
+		MonoClass *klass = t->data.klass;
+		MonoClass *eklass = m_class_get_element_class (klass);
+		MonoArray *arr = (MonoArray *) MONO_HANDLE_RAW (o);
+
+		if (m_class_get_rank (klass) > 1)
+			break;
+
+		if (arr->bounds)
```
Comment (CoffeeFlux) [inline src/mono/mono/metadata/marshal.c:6275]: Shouldn't this always be NULL for SZArrays?

File: src/mono/mono/metadata/marshal.c [hunk 4]
```diff
@@ -6264,6 +6264,31 @@ mono_marshal_asany_impl (MonoObjectHandle o, MonoMarshalNative string_encoding,
...
+			if (arr->bounds->lower_bound != 0)
+				break;
+
+		if (mono_class_is_auto_layout (eklass))
+			break;
+
+		if (m_class_is_valuetype (eklass) && (mono_class_is_explicit_layout (eklass) || m_class_is_blittable (eklass) || m_class_is_enumtype (eklass)))
+			return (arr->vector);
```
Comment (CoffeeFlux) [inline src/mono/mono/metadata/marshal.c:6283]:
```suggestion
			return arr->vector;
```

File: src/mono/mono/metadata/marshal.c [hunk 5]
```diff
@@ -6264,6 +6264,31 @@ mono_marshal_asany_impl (MonoObjectHandle o, MonoMarshalNative string_encoding,
...
+
+		if (m_class_get_rank (klass) > 1)
+			break;
+
+		if (arr->bounds)
+			if (arr->bounds->lower_bound != 0)
+				break;
+
+		if (mono_class_is_auto_layout (eklass))
+			break;
```
Comment (CoffeeFlux) [inline src/mono/mono/metadata/marshal.c:6280]: Not propagating some kind of error here feels weird to me, but we don't do it above so maybe it's fine?

File: src/mono/mono/metadata/marshal.c (hunk 4 above)
Comment (CoffeeFlux) [reply src/mono/mono/metadata/marshal.c:6283]: Also, do we really not have to copy here? What if this is marked with either `out` or `in`?

Comment (alexischr) [reply src/mono/mono/metadata/marshal.c:6283]: Yeah I think I should special-case in and out

File: src/mono/mono/metadata/marshal.c (hunk 2 above)
Comment (alexischr) [reply src/mono/mono/metadata/marshal.c:6276]: CoreCLR does the same check I believe, and there is no concept of lower bound in native arrays.

File: src/mono/mono/metadata/marshal.c (hunk 3 above)
Comment (alexischr) [reply src/mono/mono/metadata/marshal.c:6275]: Not sure, I thought bounds and sized arrays are two distinct array features and a SZArray could have bounds.

File: src/mono/mono/metadata/marshal.c (hunk 2 above)
Comment (CoffeeFlux) [reply src/mono/mono/metadata/marshal.c:6276]: Sounds good. And yeah, but you could ignore the bounds or pad with NULL bytes, so just wanted to make sure we're the same as CoreCLR.

File: src/mono/mono/metadata/marshal.c [hunk 6]
```diff
@@ -6268,6 +6268,35 @@ mono_marshal_asany_impl (MonoObjectHandle o, MonoMarshalNative string_encoding,
 
 		return res;
 	}
+	case MONO_TYPE_SZARRAY: {
```
Comment (alexischr) [inline src/mono/mono/metadata/marshal.c:6271]:
```suggestion
	case MONO_TYPE_SZARRAY: {
			//TODO: Implement structs and in-params for all value types	
```

File: src/mono/mono/metadata/marshal.c [hunk 7]
```diff
@@ -6268,6 +6268,36 @@ mono_marshal_asany_impl (MonoObjectHandle o, MonoMarshalNative string_encoding,
...
+
+		if (arr->bounds)
+			if (arr->bounds->lower_bound != 0)
+				break;
+
+		if (mono_class_is_auto_layout (eklass))
+			break;
+
+		if (m_class_is_valuetype (eklass) && (mono_class_is_explicit_layout (eklass) || m_class_is_blittable (eklass) || m_class_is_enumtype (eklass)))
+			return (arr->vector);
```
Comment (CoffeeFlux) [inline src/mono/mono/metadata/marshal.c:6292]:
```suggestion
			return arr->vector;
```

File: src/mono/mono/metadata/marshal.c [hunk 8]
```diff
@@ -6328,6 +6358,20 @@ mono_marshal_free_asany_impl (MonoObjectHandle o, gpointer ptr, MonoMarshalNativ
...
+		MonoClass *klass = t->data.klass;
+		MonoClass *eklass = m_class_get_element_class (klass);
+		MonoArray *arr = (MonoArray *) MONO_HANDLE_RAW (o);
+
+		if (eklass != mono_get_char_class ())
+			break;
+
+		mono_unichar2 *utf16_array = g_utf8_to_utf16 ((const char *)ptr, arr->max_length, NULL, NULL, NULL);
+		g_free (ptr);
+		memcpy (arr->vector, utf16_array, arr->max_length * sizeof(mono_unichar2));
```
Comment (CoffeeFlux) [inline src/mono/mono/metadata/marshal.c:6371]:
```suggestion
		memcpy (arr->vector, utf16_array, arr->max_length * sizeof (mono_unichar2));
```

**Review by CoffeeFlux — APPROVED** [review]
Just a couple nits, should be fine otherwise. I'm not the biggest fan of just returning the pointer to the managed heap but I think it's okay here with the guard against `in`. Thanks!

**Pull request merged**